### PR TITLE
[InstallAPI] Capture & compare load commands that may differ per arch slice

### PR DESCRIPTION
**Pull request by cyndyishida**
* Capture reexported libraries, allowable clients, rpaths, shared cache eligibility.
* Add support for select Xarch options.
* Add diagnostics related to capturing these options.
* Add support for verifying these attributes against what is encoded in the dylib.B

### CODE REVIEW

File: clang/lib/InstallAPI/DiagnosticBuilderWrappers.cpp [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+const DiagnosticBuilder &operator<<(const DiagnosticBuilder &DB,
+                                    const PlatformVersionSet &Platforms) {
+  std::string PlatformAsString;
+  raw_string_ostream Stream(PlatformAsString);
+
+  Stream << "[ ";
+  bool NeedsComma = false;
+  for (auto &[Platform, Version] : Platforms) {
+    if (NeedsComma)
```
Comment (zixu-w) [inline clang/lib/InstallAPI/DiagnosticBuilderWrappers.cpp:44]: `STLExtras` has an `interleaveComma(const Container &c, StreamT &os, UnaryFunctor each_fn)` that could simplify some of the trouble here.

File: clang/lib/InstallAPI/DylibVerifier.cpp [hunk 2]
```diff
@@ -702,5 +724,179 @@ DylibVerifier::Result DylibVerifier::verifyRemainingSymbols() {
...
+    return false;
+  }
+
+  if (ProvidedBA.ParentUmbrella.empty() && !DylibBA.ParentUmbrella.empty()) {
+    Ctx.Diag->Report(diag::err_parent_umbrella_missing)
+        << "installAPI option" << DylibBA.ParentUmbrella;
+    return false;
+  }
+
+  if (!ProvidedBA.ParentUmbrella.empty() && !DylibBA.ParentUmbrella.empty()) {
```
Comment (zixu-w) [inline clang/lib/InstallAPI/DylibVerifier.cpp:823]: Should this be
```
!ProvidedBA.ParentUmbrella.empty() &&  DylibBA.ParentUmbrella.empty()
                                      ^ no negation here
```
?

**Review by zixu-w — APPROVED** [review]

**Comment by cyndyishida** [comment]
Thanks for reviewing @zixu-w !

**Comment by efriedma-quic** [comment]
It looks like this is causing failures on the reverse-iteration buildbot (https://lab.llvm.org/buildbot/#/builders/54/builds/9683)

**Comment by cyndyishida** [comment]
> It looks like this is causing failures on the reverse-iteration buildbot (https://lab.llvm.org/buildbot/#/builders/54/builds/9683)

Thanks for the heads up, should be fixed by https://github.com/llvm/llvm-project/pull/89270

**Pull request merged**